test/dotnet-core-uninstall.Tests/Shared/VSVersioning/VSVersionTests.cs

```diff
@@ -0,0 +1,112 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using FluentAssertions;
+using Microsoft.DotNet.Tools.Uninstall.Shared.BundleInfo;
+using Microsoft.DotNet.Tools.Uninstall.Shared.BundleInfo.Versioning;
+using Microsoft.DotNet.Tools.Uninstall.Shared.Exceptions;
+using Microsoft.DotNet.Tools.Uninstall.Shared.VSVersioning;
+using Xunit;
+
+namespace Microsoft.DotNet.Tools.Uninstall.Tests.Shared.VSVersioning
+{
+    public class VSVersionTests
+    {
+        [Theory]
+        [InlineData(new string[] { "1.0.0" }, new bool[] { false })]
+        [InlineData(new string[] { "1.0.0", "1.0.1" }, new bool[] { true, false })]
+        [InlineData(new string[] { "2.1.0", "1.0.1" }, new bool[] { false, false })]
+        [InlineData(new string[] { "1.0.0", "1.0.1", "1.1.0" }, new bool[] { true, false, false })]
+        [InlineData(new string[] { "1.0.0", "1.0.1", "2.0.0" }, new bool[] { true, false, false })]
+        [InlineData(new string[] { "1.0.0", "1.0.1", "1.0.2" }, new bool[] { true, true, false })]
+        [InlineData(new string[] { "2.1.500", "2.1.600" }, new bool[] { false, false })]
+        [InlineData(new string[] { "2.1.500", "2.1.400", "2.1.600" }, new bool[] { false, true, false })]
+        [InlineData(new string[] { "2.2.100", "2.2.200" }, new bool[] { false, false })]
+        [InlineData(new string[] { "2.2.100", "2.2.200", "2.2.300" }, new bool[] { false, true, false })]
+        internal void TestUninstallAllowed(string[] versions, bool[] allowed)
+        {
+            var bundles = new List<Bundle<SdkVersion>>();
+            foreach (string v in versions)
+            {
+                bundles.Add(new Bundle<SdkVersion>(new SdkVersion(v), new BundleArch(), string.Empty, string.Empty));
+            }
+            bundles.ForEach(b => b.UninstallAllowed.Should().Be(true));
+
+            VisualStudioSafeVersionsExtracter.AssignUninstallAllowed(bundles);
+
+            for (int i = 0; i < versions.Length; i++)
+            {
+                bundles
+                    .Where(b => b.Version is SdkVersion)
+                    .ToArray()[i]
+                    .UninstallAllowed.Should().Be(allowed[i]);
+            }
+        }
+
+        [Theory]
+        [InlineData("1.0.0")]
+        [InlineData("2.1.0", "1.0.1" )]
+        [InlineData("2.1.500", "2.1.600")]
+        [InlineData("2.2.100", "2.2.200" )]
+        internal void TestUninstallError(params string[] versions)
+        {
+            var bundles = new List<Bundle>();
+            foreach (string v in versions)
+            {
+                bundles.Add(new Bundle<SdkVersion>(new SdkVersion(v), new BundleArch(), string.Empty, v));
+            }
+
+            VisualStudioSafeVersionsExtracter.AssignUninstallAllowed(bundles);
+            // None of the bundles are uninstallable-> throw error
+            var exception = Assert.Throws<UninstallationNotAllowedException>(() => VisualStudioSafeVersionsExtracter.RemoveUninstallableBundles(bundles));
```

Comment:
nit: Prefer using fluentassertions, here is an example of asserting exception https://github.com/dotnet/cli/blob/95c6eff6daa1a69f29c42b2d405400ad44bdec91/test/Microsoft.DotNet.ToolPackage.Tests/ToolConfigurationDeserializerTests.cs#L29